File: 3.1. Binary Search/1. Binary Search/dory.py
# Runtime 0ms
# Beats 100%
# Weird result
# Memory 17.73MB
# Beats 90.03%
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] >= target:
                if nums[mid] == target:
                    return mid
                else:
                    if nums[left] == target:
                        return left
                    elif nums[right] == target:
                        return right
                    right = mid
                    left +=1
            else:
                if nums[left] == target:
                    return left
                elif nums[right] == target:
                    return right
                else:
                    left = mid
                    right -= 1
        return -1

File: 3.1. Binary Search/1. Binary Search/dory.cs
// Runtime 0ms
// Beats 100.00%

// Memory 53.02MB
// Beats 13.64%

public class Solution {
    public int Search(int[] nums, int target) {
        int left = 0;
        int right = nums.Length - 1;
        int result = -1;
        if(nums.Length == 1 && nums[0] == target) {
            return 0;
        }
        while(left < right) {
            if(nums[left] == target) {
                return left;
            }
            if(nums[right] == target) {
                return right;
            }
            int medium = (right + left) / 2;
            if(nums[medium] == target) {
                return medium;
            } else if (nums[medium] < target) {
                left = medium + 1;
            } else {
                right = medium - 1;
            }
        }
        return result;
    }
}